chore: New doc - Onboarding Doc (#3902)
tobymao/sqlglot: posts/onboarding.md

# <Token token_type: <TokenType.EQ: 'EQ'>, text: '=', line: 1, col: 29, start: 28, end: 28, comments: []>
# <Token token_type: <TokenType.NUMBER: 'NUMBER'>, text: '1', line: 1, col: 31, start: 30, end: 30, comments: []>
```

The Tokenizer processes each of the query’s eight symbols (words or operators like `=`) by looking up each one in its mapping of keyword strings to token type.

Each token is generally represented by its `TokenType` and by its string value stored in the `text` field. Other metadata such as `line` or `col` are stored during this step and can help with error diagnostics and other tasks.

SQLGlot’s `TokenType` enum provides an indirection layer between the text of each symbol and its type. For example, `!=` and `<>` are often used interchangeably to represent "not equals", so SQLGlot groups them together by mapping them both to `TokenType.NEQ`.

The mapping of symbols to `TokenType` enum values is stored in the dictionary `Tokenizer.KEYWORDS`.

## Parser

The parser module takes the list of tokens generated by the tokenizer and builds an [abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) from them.

> The AST represents the meaning (“semantics”) of a SQL statement, loosely defined as the set of actions the statement should invoke. The commonality of semantics across databases/dialects is what allows transpilation to occur.

Reviewer: ```suggestion
The AST represents a SQL statement by storing its _meaningful_ constituent components. For example, to represent a `SELECT` query we can only store its projections, filters, aggregation expressions and so on, but the `SELECT`, `WHERE` and `GROUP BY` keywords don't need to be preserved, since they are implied.

In SQLGlot's case, the AST usually stores information that captures the _semantics_ of an expression, i.e. its meaning, which is what allows it to transpile SQL code between different dialects. A blog post that explains this idea at a high-level can [be found here](https://tobikodata.com/transpiling_sql1.html).
```

Author: This reads a bit off to me, the second paragraph seems disconnected and/or repeating the first, should we merge them?

Reviewer: Yeah, I can see how this may be confusing -- I'm open to rephrasing it.

Imo we shouldn't conflate the _semantics_ of an expression with its grammatical structure, which is what an AST is primarily meant to represent.

SQLGlot's AST is special, because we inject additional information to it as opposed to what you'd get by a typical parser. For example, some `Func` nodes store a return type, and other nodes implicily store whether the arguments can be strings or datetime values (e.g. the `TsOrDs*` family).

That's what I usually think about when I see the term "semantics". Also, keep in mind that we don't really represent the semantics accurately in many cases, e.g. when expressions are type-sensitive.

Author: I see what you mean, I think the confusing part is that I'd read `The AST represents ...` as basically `The [SQLGLot] AST represents ...` which can be conflicting with the 2nd paragraph. From what I understand your intention is to distinguish a typical AST (i.e storing just the grammatical structure) produced from a general purpose SQL parser to the "enhanced" version that is produced by SQLGlot. 

If this is the case, I think your 2nd explanation conveys the message better

Reviewer: What about:

> An AST generally stores the _meaningful_ constituent components of a statement. For example, to represent a `SELECT` query we can only store its projections, filters, aggregation expressions and so on, but the `SELECT`, `WHERE` and `GROUP BY` keywords don't need to be preserved, since they are implied.

> In SQLGlot's case, the AST usually stores additional information that captures the _semantics_ of an expression, i.e. its meaning, which is what allows it to transpile SQL code between different dialects. A blog post that explains this idea at a high-level can [be found here](https://tobikodata.com/transpiling_sql1.html).

**`_match_texts(texts)`**   | Attempt to match a keyword in a set of strings/texts

### `_parse` methods
SQLGlot’s parser follows the “recursive descent” approach, meaning that it relies on mutually recursive methods in order to understand SQL syntax and the various precedence levels between combined SQL expressions.

For example, the SQL statement `CREATE TABLE a (col1 INT)` will be parsed into an `exp.Create` expression that includes the kind of object being created (table) and its schema (table name, column names and types). The schema will be parsed into an `exp.Schema` node that contains one column definition `exp.ColumnDef` node for each column.

The relationships between SQL constructs are encoded in methods whose names begin with “_parse_”, such as `_parse_create()`. We refer to these as “parsing methods”.

For example, to parse the statement above SQLGlot calls the `_parse_create()` method. Then:
- `_parse_create()` determines that a `TABLE` is being created
- Because a table is being created, it knows the table name comes next and calls `_parse_table_parts()` (the table name may have multiple parts like `catalog.schema.table`)

> - Because a table is being created, it knows the schema definition comes after the table name and calls `_parse_schema()`

Reviewer: ```suggestion
- It also knows that a schema definition comes after the table name and calls `_parse_schema()`
```

Author: Hmm, I think if I was to change the description here I'd go with something like:

```
For example, to parse the statement above the entrypoint method `_parse_create()` is called. Then:
- In `_parse_create()`, the parser determines that a `TABLE` is being created
- Because a table is being created, the table name is expected next and thus `_parse_table_parts()` is invoked; This is because the table name may have multiple parts such as `catalog.schema.table`
- Having processed the table name, the parser continues parsing the schema definition by calling `_parse_schema()`
```

It may have a bit more noise added, but I think the bullet point breakdown is stitched together more "naturally". What do you say?

Reviewer: Sure yeah

- A schema contains comma-separated column definitions, so `_parse_schema()` calls `_parse_column_def()` on each item
- After all elements are parsed, `_parse_create()` assembles and returns an `exp.Create` expression

A key aspect of the SQLGlot parser is that the parsing methods are composable and can be passed as arguments. We describe how that works in the next section.

### Utility methods
SQLGlot provides helper methods for common parsing tasks. They should be used whenever possible to reduce code duplication and manual token/text matching.

> Many helper methods take a parsing method argument specifying the parsing method they should apply to sub-components of the object they are parsing.

Reviewer: ```suggestion
Many helper methods take a parsing method argument, which is invoked to parse part(s) of the clause they're supposed to parse.
```

Each dialect’s Generator class may specify:

- A set of flags such as `ALTER_TABLE_INCLUDE_COLUMN_KEYWORD`. As with the parser, these flags control the behavior of base Generator methods.

- Set of `Expression -> str` mappings
    - The generator traverses the AST recursively and produces the equivalent string for each expression it visits. This can be thought of as the reverse operation of the parser, where expressions / AST nodes are converted back to strings.
    - The mappings are grouped by general semantic type (e.g., data type expressions), with each type having its own dictionary.

The `Expression -> str` mappings can be implemented as either:
- Entries in the `TRANSFORMS` dictionary, which are best suited for single-line generations.
- Functions with names of the form `<expr_name>_sql(...)`. For example, to generate SQL for an `exp.SessionParameter` node we can override the base generator method by defining the method `def sessionparameter_sql(...)` in a dialect’s Generator class.

### Generating queries
The key method in the Generator module is the `sql()` function, which is responsible for generating the string representation of each expression.

> It locates the mapping of expression to string or generator `Callable` by examining  the keys of the `TRANSFORMS` dictionary and searching the `Generator`'s methods for an `<exprname>_sql()` method.

Reviewer: ```suggestion
It locates the mapping of expression to string or generator `Callable` by examining the keys of the `TRANSFORMS` dictionary and searching the `Generator`'s methods for an `<exprname>_sql()` method.
```

> It then substitutes the string or invokes the callable to return the correct SQL code.

Reviewer: ```suggestion
It then calls the appropriate callable to produce the corresponding SQL code.

> [!IMPORTANT]
> If there is both a `TRANSFORM` entry and an `<expr_name>_sql(...)` method for a given expression, the generator will use the former to convert the expression into a string.
```

The optimizer operates on the abstract syntax tree (AST) returned by the parser, transforming it into a more compact form while preserving the semantics of the original query.

> [!NOTE]
> The optimizer performs only logical optimization; The underlying engine is almost always better at optimizing for performance.

### Optimization rules
The optimizer essentially applies [a list of rules](https://sqlglot.com/sqlglot/optimizer.html) in a sequential manner, where each rule takes in an AST and produces a canonicalized and optimized version of it.

> [!WARNING]
> The rule order is important, as some rules depend on others to work properly. We discourage manually applying individual rules, which may result in erroneous or unpredictable behavior.

The first, foundational optimization task is to _standardize_ an AST, which is required by other optimization steps. These rules implement canonicalization:

#### Qualify
The most important rule in the optimizer is `qualify`, which is responsible for rewriting the AST such that all tables and columns are _normalized_ and _qualified_.

##### Identifier normalization

> The normalization step (`normalize_identifiers.py`) transforms all unquoted identifiers to lower or upper case (depending on the dialect), essentially making them case-insensitive.

Reviewer: ```suggestion
The normalization step (`normalize_identifiers.py`) transforms some identifiers, by converting them into lower or upper case, ensuring the semantics are preserved in each case (e.g. by respecting case-sensitivity).

This transformation reflects how identifiers would be resolved by the engine corresponding to each SQL dialect, and plays a very important role in the standardization of the AST.
```

Author: The explanations of these optimizer steps originate mostly from the docstrings (this one is from `normalize_identifiers.py` in particular) as I didn't want to stray off; If they're outdated or require new details to be up to date, we can also update them too (?)

Reviewer: Yep, they're probably inaccurate / outdated and we should fix them.

> > [!NOTE]
> > Some dialects, like BigQuery, treat identifiers as case-insensitive even when they're quoted, so all identifiers are normalized.

Reviewer: This isn't generally correct for BigQuery, since it treats tables (not CTEs) as case-sensitive, for example. OTOH, DuckDB does treat all identifiers as case-insensitive.

```suggestion
> [!NOTE]
> Some dialects, like DuckDB, treat identifiers as case-insensitive even when they're quoted, so all identifiers are normalized.
```

This means that all table references are assigned an alias and all column names are prefixed with their source table’s name.

This example demonstrates qualifying the query `SELECT col FROM tbl`, where `tbl` contains a single column `col`. Note that the table’s schema is passed as an argument to `qualify()`:

```Python
import sqlglot
from sqlglot.optimizer.qualify import qualify

schema = {"tbl": {"col": "INT"}}
expression = sqlglot.parse_one("SELECT col FROM tbl")
qualify(expression, schema=schema).sql()

# 'SELECT "tbl"."col" AS "col" FROM "tbl" AS "tbl"'
```

The `qualify` rule also offers a set of sub-rules that further canonicalize the query.

For instance, `qualify()` can expand stars such that each `SELECT *` is replaced by a projection list of the selected columns.

This example modifies the previous example’s query to use `SELECT *`. `qualify()` expands the star and returns the same output as before:

```Python
import sqlglot
from sqlglot.optimizer.qualify import qualify

schema = {"tbl": {"col": "INT"}}
expression = sqlglot.parse_one("SELECT * FROM tbl")
qualify(expression, schema=schema).sql()

# 'SELECT "tbl"."col" AS "col" FROM "tbl" AS "tbl"'
```

> #### Type annotation

Reviewer: ```suggestion
#### Type inference
```

Some SQL operators’ behavior changes based on the data type of its inputs.

For example, in some SQL dialects `+` can be used for either adding numbers or concatenating strings. The database uses its knowledge of column data types to determine which operation should be executed in a specific situation.

Like the database, SQLGlot needs to know column data types to perform some optimizations. In many cases, both transpilation and optimization need type information to work properly.

Type annotation begins with user-provided information about column data types for at least one table. SQLGlot then traverses the AST, propagating that type information and attempting to infer the type returned by each AST expression.

Users may need to provide column type information for multiple tables to successfully annotate all expressions in the AST.

> 
> 
> 
> 

Reviewer: Nit: let's try to be consistent about the whitespace we use. My personal preference is to always separate sections with a single blank line. Open to alternatives as long as we stick to one style.
```suggestion
```

    intermediate_table
)
SELECT
  traced_col
FROM
  cte
“””

node = lineage(
    column="traced_col",
    sql=target_query,
    sources={"intermediate_table": "SELECT * FROM root_table"},
    schema={"root_table": {"traced_col": "int"}},
)
```

The `lineage()` function’s `sql` argument takes the target query, the `sources` argument takes a dictionary of source table names and the query that produces each table, and the `schema` argument takes the column names/types of the graph’s root tables.

### Implementation details
This section describes how SQLGlot traces the lineage in the previous example.

Before lineage can be traced, the following preparatory steps must be carried out:

1. Replace the `sql` query’s table references with their `sources`’ queries, such that we have a single standalone query.

This example demonstrates how the earlier example’s reference to `intermediate_table` inside the CTE is replaced with the query that generated it, `SELECT * FROM root_table`, and given the alias `intermediate_table`:

```SQL
WITH cte AS (
  SELECT
    traced_col
  FROM (
    SELECT
      *
    FROM
      root_table
  ) AS intermediate_table
)
SELECT
  traced_col
FROM
  cte
```

> 2. Qualify the standalone query, expanding `*`s and annotating all column names with source tables:

Reviewer: ```suggestion
2. Qualify the query produced by the earlier step, expanding `*`s and qualifying all column references with the corresponding source tables:
```

A `MappingSchema` object is defined with a nested dictionary, as in this example:

```Python
schema = MappingSchema({"t": {"x": "int", "y": "datetime"}})
```

The keys of the top-level dictionary are table names (`t`), the keys of `t`’s dictionary are its column names (`x` and `y`), and the values are the column data types (`int` and `datetime`).

A schema provides information required by other SQLGlot modules (most importantly, the optimizer and column level lineage).

Schema information is used to enrich ASTs with actions like replacing stars (`SELECT *`) with the names of the columns being selected from the upstream tables.

## Optimizer

The optimizer module in SQLGlot (`optimizer.py`) is responsible for producing canonicalized and efficient SQL queries by applying a series of optimization rules.

Optimizations include simplifying expressions, removing redundant operations, and rewriting queries for better performance.

The optimizer operates on the abstract syntax tree (AST) returned by the parser, transforming it into a more compact form while preserving the semantics of the original query.

> [!NOTE]

> > The optimizer does not currently support statistics, so it cannot employ cost-based optimizations such as join reordering

Reviewer 2: It’s more than just stats, right?

sqlglot only does _logical_ optimization. 

Maybe we should make a note that the underlying engine is almost always better at optimizing for performance.

Author: Right, my (inverse) thought was that physical optimization is basically cost-based optimization which requires stats to do properly, but we can simply state that we only do logical optimization

Reviewer 2: Technically, you could collect stats and write a join ordering rule in sqlglot.

So I think logical vs physical is a more precise distinction. Or in other words, sqlglot’s optimizer operates on the AST, not query plans, which is a fundamental limitation.

>

### Optimization rules
The optimizer essentially applies [a list of rules](https://sqlglot.com/sqlglot/optimizer.html) in a sequential manner, where each rule takes in an AST and produces a canonicalized and optimized version of it.

> [!WARNING]
> The rule order is important, as some rules depend on others to work properly. We discourage manually applying individual rules, which may result in erroneous or unpredictable behavior.

> The first, foundational optimization task is to _canonicalize_ an AST, which is required by other optimization steps. These rules implement canonicalization:

Author: Hmm right, the terms "canonicalize", "normalize" etc are used in various places around SQLGlot so it can definitely be confusing.. Would "standardize" be the next best verb here?
